Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Sharing/MarkerVisualDetectorSpatialLocalizer.cs

```diff
@@ -0,0 +1,179 @@
+
+#if UNITY_EDITOR
+        private void OnValidate()
+        {
+            FieldHelper.ValidateType<IMarkerDetector>(MarkerDetector);
+        }
+#endif
+
+        /// <inheritdoc />
+        public override bool TryCreateLocalizationSession(IPeerConnection peerConnection, MarkerVisualDetectorLocalizationSettings settings, out ISpatialLocalizationSession session)
+        {
+            markerDetector = (markerDetector == null) ? MarkerDetector as IMarkerDetector : markerDetector;
+            session = new LocalizationSession(this, settings, peerConnection, debugLogging);
+            return true;
+        }
+
+        /// <inheritdoc />
+        public override bool TryDeserializeSettings(BinaryReader reader, out MarkerVisualDetectorLocalizationSettings settings)
+        {
+            settings = new MarkerVisualDetectorLocalizationSettings();
+            return true;
+        }
+
+        private class LocalizationSession : DisposableBase, ISpatialLocalizationSession
+        {
+            private readonly MarkerVisualDetectorSpatialLocalizer localizer;
+            private readonly MarkerVisualDetectorLocalizationSettings settings;
+            private readonly IPeerConnection peerConnection;
+            private readonly ISpatialCoordinateService coordinateService;
+            private readonly bool debugLogging = false;
+            private readonly TaskCompletionSource<string> coordinateAssigned = null;
+            private readonly CancellationTokenSource discoveryCTS = null;
+
+            private string coordinateId = string.Empty;
+
+            public LocalizationSession(MarkerVisualDetectorSpatialLocalizer localizer, MarkerVisualDetectorLocalizationSettings settings, IPeerConnection peerConnection, bool debugLogging = false)
+            {
+                DebugLog("Session created");
+                this.localizer = localizer;
+                this.settings = settings;
+                this.peerConnection = peerConnection;
+                this.debugLogging = debugLogging;
+
+                this.coordinateAssigned = new TaskCompletionSource<string>();
+                this.coordinateService = new MarkerDetectorCoordinateService(this.localizer.markerDetector, debugLogging);
+                this.discoveryCTS = new CancellationTokenSource();
+            }
+
+            /// <inheritdoc />
+            protected override void OnManagedDispose()
+            {
+                coordinateService.Dispose();
+                discoveryCTS.Dispose();
+            }
+
+            /// <inheritdoc />
+            public async Task<ISpatialCoordinate> LocalizeAsync(CancellationToken cancellationToken)
+            {
+                DebugLog($"Waiting for marker visual, CanBeCanceled:{cancellationToken.CanBeCanceled}, IsCancellationRequested:{cancellationToken.IsCancellationRequested}");
+                await Task.WhenAny(coordinateAssigned.Task, Task.Delay(-1, cancellationToken));
```

Comment:
>await Task.WhenAny(coordinateAssigned.Task, Task.Delay(-1, cancellationToken)); [](start = 16, length = 79)

There's a helper extension method Unless(cancellationToken) that makes this pattern a little easier to read. It does basically this same thing, but you see something like "await coordinateAssigned.Task.Unless(cancellationToken)"